src/Voron/Data/CompactTrees/CompactTree.Dictionaries.cs

```diff
@@ -83,7 +85,7 @@ public bool TryImproveDictionaryByFullScanning()
             _state.TreeDictionaryId = newDictionary.PageNumber;
 
         // We will update the number of entries regardless if we updated the current dictionary or not. 
-        _state.NextTrainAt = (long)(_state.NextTrainAt * 1.5);
+        _state.NextTrainAt = (long)(Math.Max(_state.NextTrainAt, _state.NumberOfEntries) * 1.5);
```

Comment:
I don't like this. The issue is that you are counting on the _number of entries_ here, and not the number of changes.
Consider the pretty common case of an index that has a lot of churn (documents coming & going).
The total number of writes is huge, but the number of entries is pretty stable.

Reply:
This function monotonically growing in a stepwise way. While those cases may happen, they can be scheduled manually or in a maintenance task. This is the default growing heuristics. I am not that sure number of changes would actually improve anything. What I am saying is that statistically, unless the entries have a very different distribution than the original dictionary, after less than 6 iterations it would be difficult to improve upon the dictionary.

Reply:
Problem is, you are _never_ going to have a LOT of entries in the tree.
Consider a `NextTrainAt` of 25K - but there are never enough terms, but the value is something like the date, which changes over time.
So we want to trigger by number of changes, not total number of records.

Reply:
Question that comes after that, if we dont have enough entries why do we care anyways :D ... those are going to be very small trees even with a suboptimal encoding. Remember each dictionary (at least for now) is roughly 8 pages.

Reply:
That isn't what your code is doing, however.
For small trees, we don't care.

But for bigger trees? Consider an index that has a stable total size of ~10M, but entries keep being replaced with new data.
We certain want to train there, no?

Reply:
You can trigger it as part of a maintenance task, but I dont think we should bake that decision in Corax itself. This is something that the user would know better. The good thing about this is that we can request whenever we want to, we are not limited to the decisions Corax make. It makes sense for Corax itself tracking growth heuristics, but usage heuristics like that are better served by upper layers knowing when to request another evaluation.